### PR TITLE
Parse legacy shared code content on remote load

### PR DESCRIPTION
**Pull request by webrouse**
Part of: https://keboola.atlassian.net/browse/COM-1138

### CODE REVIEW

File: internal/pkg/mapper/sharedcode/codes/load_test.go [hunk 1]
```diff
@@ -10,6 +10,23 @@ import (
 	"github.com/keboola/keboola-as-code/internal/pkg/model"
 )
 
+func TestSharedCodeLegacyRemoteCodeContent(t *testing.T) {
+	t.Parallel()
+	context, row, _ := createTestFixtures(t, `keboola.snowflake-transformation`)
+	row.Content.Set(model.SharedCodeContentKey, "SELECT 1; \n  SELECT 2; \n ")
```
Comment (webrouse) [inline internal/pkg/mapper/sharedcode/codes/load_test.go:16]: Tak este jeden fix.
Legacy shared code content, ktory je `string`, a nie pole .... moze obsahovat nerozdelene SQL queries.

File: internal/pkg/mapper/sharedcode/codes/load_test.go [hunk 2]
```diff
@@ -10,6 +10,23 @@ import (
...
+	event := model.OnObjectsLoadEvent{
+		StateType:  model.StateTypeRemote,
+		NewObjects: []model.Object{row},
+		AllObjects: context.State.RemoteObjects(),
+	}
+	assert.NoError(t, NewMapper(context).OnObjectsLoad(event))
+
+	v, found := row.Content.Get(model.SharedCodeContentKey)
+	assert.True(t, found)
+	assert.Equal(t, []interface{}{"SELECT 1;", "SELECT 2;"}, v)
```
Comment (webrouse) [inline internal/pkg/mapper/sharedcode/codes/load_test.go:27]: Pri nacitani ich musime parsovat.

File: internal/pkg/mapper/sharedcode/codes/load.go [hunk 3]
```diff
@@ -14,6 +14,57 @@ type loader struct {
 	errors    *utils.Error
 }
 
+// OnObjectsLoad converts legacy "code_content" string -> []interface{}.
+func (m *mapper) OnObjectsLoad(event model.OnObjectsLoadEvent) error {
```
Comment (webrouse) [inline internal/pkg/mapper/sharedcode/codes/load.go:18]: Implementacia.

File: internal/pkg/mapper/sharedcode/codes/save.go [hunk 4]
```diff
@@ -56,24 +56,25 @@ func (w *writer) save() error {
...
-	} else if codeContentSlice, ok := raw.([]interface{}); ok {
-		var scripts []string
-		for _, script := range codeContentSlice {
-			scripts = append(scripts, cast.ToString(script))
-		}
-		codeContent = strhelper.TransformationScriptsToString(scripts, targetComponentId)
-	} else {
-		return fmt.Errorf(`key "%s" must be string or string[], found %T, in %s`, model.ShareCodeContentKey, raw, w.configRow.Desc())
+	// Content must be []interface{}
+	codeContentSlice, ok := raw.([]interface{})
```
Comment (webrouse) [inline internal/pkg/mapper/sharedcode/codes/save.go:60]: Kedze sa to riesi uz pri `remote load`, a uz to bolo vyriesene pre `local load`, tak tu uz kontrola byt nemusi. ... ocakavame iba `[]interface{}`.

File: internal/pkg/mapper/sharedcode/codes/save.go [hunk 5]
```diff
@@ -56,24 +56,25 @@ func (w *writer) save() error {
...
+	if !ok {
+		return fmt.Errorf(`key "%s" must be array, found %T, in %s`, model.SharedCodeContentKey, raw, w.configRow.Desc())
 	}
 
 	// Remove code content from JSON
-	rowContent.Delete(model.ShareCodeContentKey)
+	rowContent.Delete(model.SharedCodeContentKey)
+
+	// Convert []interface{} -> []string
+	var scripts []string
```
Comment (webrouse) [inline internal/pkg/mapper/sharedcode/codes/save.go:69]: Presunute z kodu vyssie.

File: internal/pkg/mapper/sharedcode/codes/save_test.go [hunk 6]
```diff
@@ -22,39 +22,15 @@ func TestSharedCodeSaveMissingKey(t *testing.T) {
 	assert.Len(t, recipe.ExtraFiles, 0)
 }
 
-func TestSharedCodeSaveString(t *testing.T) {
```
Comment (webrouse) [inline internal/pkg/mapper/sharedcode/codes/save_test.go:25]: `TestSharedCodeSaveString` test zmazany, riesi sa to pri load ... tj. interna reprezentacia uz nemoze byt `string`.

File: internal/pkg/mapper/transformation/load.go [hunk 7]
```diff
@@ -60,6 +60,9 @@ func (m *transformationMapper) MapAfterRemoteLoad(recipe *model.RemoteLoadRecipe
 			code.ConfigId = config.Id
 			code.BlockIndex = block.Index
 			code.Index = codeIndex
+			for i, script := range code.Scripts {
+				code.Scripts[i] = strhelper.NormalizeScript(script)
```
Comment (webrouse) [inline internal/pkg/mapper/transformation/load.go:64]: + este jedna drobnost ... v transformacii, v API, mozu jednotlive `scripts` obsahovat na konci biele znaky ... pri nacitani z `local` ich odstranujeme (lebo ich moze doplnit IDE), .... ale to musime spravit aj pri `remote` datach, ... preto uzivatel moze zadat prazdne znaky na koniec aj v UI.

File: internal/pkg/diff/reporter.go [hunk 8]
```diff
@@ -200,9 +200,31 @@ func (r *Reporter) objectPath(value reflect.Value) string {
...
+	if local.IsValid() {
+		localType = local.Type()
+		if localType.Kind().String() == `interface` {
+			local = local.Elem()
+			localType = local.Type()
+		}
+	}
+
+	// Print types if differs
+	includeType := remote.IsValid() && local.IsValid() && remoteType.String() != localType.String()
```
Comment (webrouse) [inline internal/pkg/diff/reporter.go:223]: Toto tu nefungovalo vzdy spravne ... ak su hodnoty `interface`, tak sa ich typy `=`, ... musime najprv previest `interface` na hodnotu vid vyssie ^^^^.

File: internal/pkg/diff/reporter_test.go [hunk 9]
```diff
@@ -27,6 +27,15 @@ func TestReporterValuesDiffSameType2(t *testing.T) {
 	}, out)
 }
 
+func TestReporterValuesDiffSameTypeInterface(t *testing.T) {
+	t.Parallel()
+	out := valuesDiff(reflect.ValueOf(interface{}([]int{1, 2})), reflect.ValueOf(interface{}([]int{3, 4})))
```
Comment (webrouse) [inline internal/pkg/diff/reporter_test.go:32]: Test, `interfaces`, rovnaky typ.

File: internal/pkg/diff/reporter_test.go [hunk 10]
```diff
@@ -45,6 +54,15 @@ func TestReporterValuesDiffDifferentType2(t *testing.T) {
 	}, out)
 }
 
+func TestReporterValuesDiffDifferentTypeInterface(t *testing.T) {
+	t.Parallel()
+	out := valuesDiff(reflect.ValueOf(interface{}([]float64{1, 2})), reflect.ValueOf(interface{}([]int{1, 2})))
```
Comment (webrouse) [inline internal/pkg/diff/reporter_test.go:59]: Test, `interfaces`, rozny typ.

**Review by webrouse — COMMENTED** [review]
`Shared Config` uz viem nacitat spravne:
image

Takze vacsina edge-cases by mala byt osetrena.
Toto by mal byt posledny.

**Review by JakubMatejka — APPROVED** [review]
